kotlin-analysis-api/src/main/kotlin/com/google/devtools/ksp/impl/symbol/kotlin/KSClassifierReferenceImpl.kt

```diff
@@ -1,37 +1,77 @@
 package com.google.devtools.ksp.impl.symbol.kotlin
 
 import com.google.devtools.ksp.IdKeyPair
+import com.google.devtools.ksp.IdKeyTriple
 import com.google.devtools.ksp.KSObjectCache
 import com.google.devtools.ksp.symbol.*
-import org.jetbrains.kotlin.analysis.api.components.buildClassType
-import org.jetbrains.kotlin.analysis.api.types.KtUsualClassType
+import org.jetbrains.kotlin.analysis.api.types.KtClassType
+import org.jetbrains.kotlin.analysis.api.types.KtClassTypeQualifier
+import org.jetbrains.kotlin.analysis.api.types.KtTypeParameterType
 
 class KSClassifierReferenceImpl private constructor(
-    internal val ktType: KtUsualClassType,
+    internal val ktType: KtClassType,
+    internal val index: Int,
     override val parent: KSTypeReference?
 ) : KSClassifierReference {
-    companion object : KSObjectCache<IdKeyPair<KtUsualClassType, KSTypeReference?>, KSClassifierReferenceImpl>() {
-        fun getCached(ktType: KtUsualClassType, parent: KSTypeReference?) =
-            cache.getOrPut(IdKeyPair(ktType, parent)) { KSClassifierReferenceImpl(ktType, parent) }
+    companion object : KSObjectCache<IdKeyTriple<KtClassType, Int, KSTypeReference?>, KSClassifierReferenceImpl>() {
+        fun getCached(ktType: KtClassType, parent: KSTypeReference?, index: Int) =
+            cache.getOrPut(IdKeyTriple(ktType, index, parent)) { KSClassifierReferenceImpl(ktType, index, parent) }
     }
+
+    private val classifierReference: KtClassTypeQualifier
+        get() = ktType.qualifiers[index]
+
     override val qualifier: KSClassifierReference? by lazy {
-        ktType.classId.outerClassId?.let {
-            analyze {
-                buildClassType(it)
-            }
-        }?.let { getCached(it as KtUsualClassType, parent) }
+        if (index == 0) {
+            null
+        } else {
+            getCached(ktType, parent, index - 1)
+        }
     }
 
     override fun referencedName(): String {
-        return ktType.classId.asFqNameString()
+        return classifierReference.name.asString()
     }
 
     override val typeArguments: List<KSTypeArgument> by lazy {
-        ktType.typeArguments().map { KSTypeArgumentImpl.getCached(it, this) }
+        classifierReference.typeArguments.map { KSTypeArgumentImpl.getCached(it, this) }
     }
 
     override val origin: Origin = parent?.origin ?: Origin.SYNTHETIC
 
     override val location: Location
         get() = parent?.location ?: NonExistLocation
+
+    override fun toString(): String {
+        return referencedName()
+    }
+}
+
+class KSClassifierParameterImpl private constructor(
```

Comment:
At a glance, everything else is in its own file, but why not this one?

Reply:
My thought on this is that this implementation still is a classifier reference element implementation, in FE 1.0 this is not separated out because it was PSI based, therefore I want to keep logically same reference elements in same file.